### PR TITLE
refactor all positions

### PR DESCRIPTION
**Pull request by catalinnnn**
Basic tests for important flows.
Basic functionality for important flows.

Made partial repay possible.
Refactor positions and simplified burn/mint flows.

Still need to work on liquidation flow, more things need to be discussed&done, but on other PRs.

### CODE REVIEW

File: lending_pool/src/lib.rs [hunk 1]
```diff
@@ -89,32 +88,16 @@ pub trait LendingPool:
         self.require_non_zero_address(&initial_caller)?;
 
         let borrow_token_pool_address = self.get_pool_address_non_zero(&asset_to_borrow)?;
-        let lend_token_pool_address = self.get_pool_address_non_zero(&payment_lend_id)?;
+        let ltv = self.get_ltv_exists_and_non_zero(&collateral_token_id)?;
```
Comment (dorin-iancu) [inline lending_pool/src/lib.rs:91]: I'd rename this variable to `loan_to_value`.

File: liquidity_pool/src/liquidity.rs [hunk 2]
```diff
@@ -234,75 +221,66 @@ pub trait LiquidityModule:
                 .direct(&initial_caller, asset_token_id, 0, &extra_asset_paid, &[]);
         }
 
-        // TODO: Instead of borrow_token_amount (i.e. 1:1 ratio), calculate how much collateral amount was repaid
```
Comment (dorin-iancu) [inline liquidity_pool/src/liquidity.rs:237]: Keep this TODO, as it seems to me like the ratio is still 1:1

File: liquidity_pool/src/liquidity.rs [hunk 3]
```diff
@@ -234,75 +221,66 @@ pub trait LiquidityModule:
...
-                .debt_positions()
-                .insert(debt_position_id.clone(), debt_position.clone());
+            self.borrow_position(borrow_token_nonce)
+                .set(&borrow_position);
         }
 
         self.send()
             .esdt_local_burn(borrow_token_id, borrow_token_nonce, borrow_token_amount);
 
-        // Same here, use calculated amount of repaid collateral instead of borrow_token_amount
```
Comment (dorin-iancu) [inline liquidity_pool/src/liquidity.rs:250]: Keep TODO

File: liquidity_pool/src/liquidity.rs [hunk 4]
```diff
@@ -234,75 +221,66 @@ pub trait LiquidityModule:
...
-        );
-        require!(
-            !debt_position.is_liquidated,
-            "position is already liquidated"
-        );
         require!(
-            debt_position.health_factor < self.health_factor_threshold().get(),
-            "the health factor is not low enough"
+            !self.borrow_position(borrow_position_nonce).is_empty(),
+            "Position is empty"
```
Comment (dorin-iancu) [inline liquidity_pool/src/liquidity.rs:265]: Maybe change `empty` to `repaid or already liquidated`?

File: safety_module/src/lib.rs [hunk 5]
```diff
@@ -254,16 +253,21 @@ pub trait SafetyModule {
...
-            deposit_metadata,
-            &[BoxedBytes::empty()],
-        );
+    fn mint_deposit_nft(
+        self,
+        deposit_metadata: &DepositPosition<Self::BigUint>,
+        amount: Self::BigUint,
+    ) {
+        self.send()
+            .esdt_nft_create::<DepositPosition<Self::BigUint>>(
```
Comment (dorin-iancu) [inline safety_module/src/lib.rs:262]: No need to have the type here, the Rust compiler can deduce it.
```suggestion
            .esdt_nft_create(
```

File: lending_pool/src/lib.rs (hunk 1 above)
Comment (catalinnnn) [reply lending_pool/src/lib.rs]: done

File: liquidity_pool/src/liquidity.rs (hunk 2 above)
Comment (catalinnnn) [reply liquidity_pool/src/liquidity.rs:237]: done

File: liquidity_pool/src/liquidity.rs (hunk 3 above)
Comment (catalinnnn) [reply liquidity_pool/src/liquidity.rs:250]: done

File: liquidity_pool/src/liquidity.rs (hunk 4 above)
Comment (catalinnnn) [reply liquidity_pool/src/liquidity.rs]: done

File: safety_module/src/lib.rs (hunk 5 above)
Comment (catalinnnn) [reply safety_module/src/lib.rs]: done

**Review by mateigraura — APPROVED** [review]

**Pull request merged**